src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNISslStream.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Security;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    /// <summary>
+    /// This class extends SslStream to customize stream behavior for Managed SNI implementation.
+    /// </summary>
+    internal class SNISslStream : SslStream
+    {
+        private readonly ConcurrentQueueSemaphore _writeAsyncQueueSemaphore;
+        private readonly ConcurrentQueueSemaphore _readAsyncQueueSemaphore;
+
+        public SNISslStream(Stream innerStream, bool leaveInnerStreamOpen, RemoteCertificateValidationCallback userCertificateValidationCallback)
+            : base(innerStream, leaveInnerStreamOpen, userCertificateValidationCallback)
+        {
+            _writeAsyncQueueSemaphore = new ConcurrentQueueSemaphore(1);
+            _readAsyncQueueSemaphore = new ConcurrentQueueSemaphore(1);
+        }
+
+        // Prevent the ReadAsync's collision by running task in Semaphore Slim
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return _readAsyncQueueSemaphore.WaitAsync().ContinueWith<int>(t => base.ReadAsync(buffer, offset, count, cancellationToken).Result)
```

Comment:
Although this no longer blocks synchronously on the semaphore, calling Task.Result still blocks synchronously on ReadAsync (sync-over-async, with the associated TP starvation issues).

Is there any particular reason to avoid async/await here, i.e. something like the following:

```c#
public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
{
    await _readAsyncQueueSemaphore.WaitAsync().ConfigureAwait(false);
    try {
        await base.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
    } finally {
        _readAsyncQueueSemaphore.Release(t.Result);
    }
}
```

_readAsyncQueueSemaphore.WaitAsync could even be changed to return an IDisposable which calls Release, in which case the above could be simplified with `using`.

Reply:
I'm learning :) sure, will update it!

Reply:
We await on these async methods (ReadAsync/WriteAsync) in SslOverTdsStream class where `Stream` is actually in use. Since this class is mainly an extension to these APIs to ensure we don't execute simultaneous read/writes, we don't await here. We also don't want this class to be blocking (correction to my last statement) and only want calls to be sent in queue 1 at a time.

I also mentioned in another thread/issue, can repost it here. I think this could have been handled by `SslStream` class itself. This design of Semaphore Slim + Concurrent Queue on SslStream creates an extended class `SNISslStream` that we will now base our Stream on. It's very much an independent design, and maybe something for System.Net.Security to consider ? Since like us, many other clients can land into concurrency issues with this class (as per their documentation of limited support).

Have you come across similar case in Ngpsql?

Reply:
> We await on these async methods (ReadAsync/WriteAsync) in SslOverTdsStream class where Stream is actually in use. Since this class is mainly an extension to these APIs to ensure we don't execute simultaneous read/writes, we don't await here.

OK. But note that implementing the above ReadAsync wrapper with async/await or with ContinueWith will in the end produce exactly the same method - async/await is precisely about breaking the method up and wiring the different parts with callbacks. In the end, both produce a method that doesn't block and returns a Task.

In other words, async/await is just a simpler and safer way to express the same thing you're doing with ContinueWith (since the compiler is helping you out).

Re concurrent usage of streams... I know that NetworkStream does internally serialize concurrent requests via a queue, and that it's a think that is very much regretted nowadays, as the queue adds some significant (mandatory!) overhead for all network operations. At the end of the day, sockets cannot be used concurrently, and like any other non-thread-safe resource, it should IMHO be up to the application to perform any necessary synchronization to avoid concurrent usage. So I would be surprised if the BCL folk would agree to introducing a serializing queue mechanism inside SslStream...

I'm not sure what the exact requirements of SqlClient are, but in Npgsql locking/synchronization happen at a higher level than the stream. For example, I think SqlClient already employs high-level synchronization to prevent multiple commands to be executed concurrently on the same connection - that's the kind of stuff that Npgsql does as well. I'm not sure what other cases you have to deal with, when multiple writes are attempted concurrently...

Reply:
> In other words, async/await is just a simpler and safer way to express the same thing you're doing with ContinueWith (since the compiler is helping you out).

I couldn't get async/await to fix things as ContinueWith does.. Maybe there're more areas in Managed SNI we need to scan and fix. I'll make a note of revisiting that when we do improvements in future. For now, we're in crunch for release so we'll let it stay ContinueWith since that makes a whole lot of difference.

> I'm not sure what the exact requirements of SqlClient are, but in Npgsql locking/synchronization happen at a higher level than the stream. For example, I think SqlClient already employs high-level synchronization to prevent multiple commands to be executed concurrently on the same connection - that's the kind of stuff that Npgsql does as well. I'm not sure what other cases you have to deal with, when multiple writes are attempted concurrently...

We have a whole lot of issues due to concurrency and I suspect the higher level of synchronization is broken somewhere. We definitely need to redesign and make things consistent in Managed SNI as it's not a clean design right now. This time we're going to fix issues for customers who are blocked with huge problems that are getting fixed here. We'll revisit this and try to set things straight! Thanks again for your thoughts and ideas, highly appreciated! :)

Reply:
Sure thing @cheenamalhotra. I'd still highly recommend not blocking the thread - the part where Task.Result is called - but putting a ContinueWith on it instead to make it truly async. Beyond that, async/await would improve code quality (and likely also perf), but both patterns are correct bug-wise.

Reply:
I can see why async/await was needed as we tested more and more cases (changed it).
But implementing async/awaits brings back #422 and Readers go to blocking. I'm continuing to look further.

Reply:
We're keeping the implementation as it is (using async/await) for now, as we need to dig more into other 2 issues.

I'll put some notes what we're facing issue:

Basically in #422 repro, we have parallel execution of sync APIs: `Connection.Open()` followed by `SqlCommand.ExecuteReader()` and when MARS is enabled, the driver only reads MARS packet asynchronously during `Connection.Open()`, rest data is read synchronously. We are suspecting these `ReadAsync()` calls are blocking the second sync API `ExecuteReader()` to read data as every other thread is simply waiting for packet to arrive that shall contain metadata information.

When I used the ContinueWith pattern above, that unblocked everything and I think that happened because these calls became synchronous. It also fixed #659 that also is on the same path. Let us know if you have ideas/tips to what we might be dealing with here, whether it's threadpool starvation or sync-async collision.. there're also so many locks in the code that are suspicious.

Reply:
@Wraith2 

Hope you're doing well! 
As you have much more experience in Managed SNI design, please feel free to chime in and share your thoughts 🙂

Reply:
I don't know about "design", I just fiddle with the existing implementation more than is healthy.

We should use `ConfigureAwait(false)` everywhere and then root cause any bugs that it exposes. The intention is that we do not care about which thread continuations are called on and those bugs are evidence that something somewhere relies on something we don't want.

If the semaphore slim is analogous to a multithread/task aware lock then it's a reasonable thing to do to replace the `lock` keyword usage with it. Otherwise I'd have to do some digging. 

So far I've only seen one reason for multiple send operations to be in flight at once and that's cancellation of a running query. What sort of queue depth are you expecting to see here and is it possible to use less expensive primitive to optimize for the 90% case of single packets?

On the idea of returning a disposable instead of using `try{}finally{}` I agree with the code purity but it has a perf penalty because the disposable will now be heap allocated as it becomes part of the closure, so I'd probably decide against it given how low level and frequently called this path is.

Reply:
> We should use ConfigureAwait(false) everywhere and then root cause any bugs that it exposes.

Absolutely.

Note that the latest compiler versions come with a warning for missing ConfigureAwait ([CA2007](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2007)), which can make your build fail if you forget one. This can be enabled as an error in an .editorconfig [as here](https://github.com/dotnet/efcore/blob/main/src/.editorconfig#L48), and you'd probably also need to set `<AnalysisLevel>latest</AnalysisLevel>` in your csproj.

> On the idea of returning a disposable instead of using try{}finally{} I agree with the code purity but it has a perf penalty because the disposable will now be heap allocated as it becomes part of the closure, so I'd probably decide against it given how low level and frequently called this path is.

Where is there a closure here?

Note also that if the disposable is a readonly struct, there's no heap allocation (as long as the struct type is being passed around and never cast to IDisposable). So this is a zero-allocation mechanism used frequently in high-perf scenarios ([see this example in EF Core](https://github.com/dotnet/efcore/blob/main/src/EFCore/Infrastructure/ConcurrencyDetectorCriticalSectionDisposer.cs)).

Reply:
The specific case we're talking about is the async method, because it's async the compiler generated object for params and locals is the closure I'm refering to, perhaps I'm using incorrect terminology. Even if you use a struct it'll end up in the state object and if it goes async it'll get boxed. You can't use ref structs in an async method so that method isn't possible in this case.

Reply:
Thanks for the details.
I will continue on these issues and keep eye on async/await to happen everywhere and see if we can remove locks and come to SemaphoreSlim to control entry if needed.

I just realized [SemaphoreSlim ](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim) is concurrent, so Queue can be removed and design can be simplified (learning 🙃)